File: spring-boot-sample-common/src/main/java/com/github/jccode/springbootsample/core/service/CrudServiceImpl.java
package com.github.jccode.springbootsample.core.service;

import com.github.jccode.springbootsample.core.repo.CrudMapper;
import com.github.jccode.springbootsample.core.model.BaseEntity;

import java.time.chrono.ChronoLocalDateTime;
import java.util.Date;

public class CrudServiceImpl<T extends BaseEntity> implements CrudService<T> {

    private final CrudMapper<T> mapper;

    public CrudServiceImpl(CrudMapper<T> mapper) {
        this.mapper = mapper;
    }

    public T find(Integer id) {
        return mapper.selectByPrimaryKey(id);
    }

    public int save(T t) {
        Integer id = t.getId();
        int ret;
        if (id == null || id == 0) {
            Date now = new Date();
            t.setCreateTime(now);
            t.setUpdateTime(now);
            ret = mapper.insert(t);
        } else {
            if (t.getUpdateTime() == null) {
                t.setUpdateTime(new Date());
            }
            ret = mapper.updateByPrimaryKey(t);
        }
        return ret;
    }

    public int updateSelective(T t) {
        return mapper.updateByPrimaryKeySelective(t);
    }

    public int delete(Integer id) {
        return mapper.deleteByPrimaryKey(id);
    }
}

File: spring-boot-sample-common/src/main/scala/com/github/jccode/springbootsample/core/service/CrudServiceImpl.scala
package com.github.jccode.springbootsample.core.service

import java.util.Date

import com.github.jccode.springbootsample.core.model.BaseEntity
import com.github.jccode.springbootsample.core.repo.CrudMapper

class CrudServiceImpl[T <: BaseEntity](val mapper: CrudMapper[T]) extends CrudService[T] {
  override def find(id: Int): T = mapper.selectByPrimaryKey(id)

  override def save(t: T): Int = {
    if (t == null || t.id == 0) {
      val now = new Date()
      t.createTime = now
      t.updateTime = now
      mapper.insert(t)
    } else {
      if (t.updateTime == null) {
        t.updateTime = new Date()
      }
      mapper.updateByPrimaryKey(t)
    }
  }

  override def updateSelective(t: T): Int = mapper.updateByPrimaryKeySelective(t)

  override def delete(id: Integer): Int = mapper.deleteByPrimaryKey(id)
}
